esp-hal/src/lib.rs

```diff
@@ -644,3 +635,18 @@ macro_rules! before_snippet {
         ))
     };
 }
+
+use crate::{
+    clock::{ClockControl, Clocks, CpuClock},
+    peripherals::Peripherals,
+};
+
+/// Initialize the system.
+///
+/// This function sets up the CPU clock and returns the peripherals and clocks.
+pub fn init(clock_config: CpuClock) -> (Peripherals, Clocks<'static>) {
```

Comment:
I think if we're going to make a radical change, we should do it in a way that allows us to extend this without a bunch of breaking changes. For instance, we may want to configure other parameters, clocks, behaviours etc.

I suggest a global `#[non_exhaustive]` config, which for now can just contain CPU clock configuration.